### PR TITLE
Milestore 1 : Small gradual type checker

### PR DESCRIPTION
**Pull request by migeed-z**
Milestone 1 - MVP of type migration for toy programs (e.g. ~3 operations)

Operations are: reshape, transpose, add

### CODE REVIEW

**Comment by jamesr66a** [comment]
Same comment as the other PR. Please use [ghstack](https://github.com/ezyang/ghstack) to make this easier to review!

File: torch/fx/experimental/gradual_graph_typechecker.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            t1 = type_check_node(n.args[0], env, expr_env)
+            t2 = type_check_node(n.args[1], env, expr_env)
+            if consistency(t1, t2):
+                expr_env[n.name] = t1
+                n.type = t1
+                return n.type
+            else:
+                return False
+
+        if n.name == 'reshape':
```
Comment (jamesr66a) [inline torch/fx/experimental/gradual_graph_typechecker.py:59]: Names are not load-bearing, they're only used for codegen and debugging. What you want to check is the target, e.g. `if n.target == torch.reshape`

File: torch/fx/experimental/gradual_graph_typechecker.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            t1 = type_check_node(n.args[0], env, expr_env)
+            t2 = n.args[1]
+            prod1 = reduce(lambda x, y: x*y, t1.__args__)
+            prod2 = reduce(lambda x, y: x*y, t2)
+            if prod1 == prod2:
```
Comment (jamesr66a) [inline torch/fx/experimental/gradual_graph_typechecker.py:64]: I think there needs to be special handling of when the shape array has a `-1` value in it

File: torch/fx/experimental/gradual_graph_typechecker.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    if n.op == 'placeholder':
+        if n.name in env.keys():
+            n.type = env[n.name]
+            return n.type
+        else:
+            raise Exception("free variable")
+
+    if n.op == 'call_function':
+        if n.name == 'add':
+            t1 = type_check_node(n.args[0], env, expr_env)
```
Comment (jamesr66a) [inline torch/fx/experimental/gradual_graph_typechecker.py:50]: Can you explain the control flow here? We're already applying `type_check_node` on every node in `type_check`, but then we're recursively calling it as well on all of the argument nodes for each of these rules. At the very least this is redundant iiuc, unless this is side-effectful?

Is it sufficient to just type check the nodes in the topological order of the program? or do we need this recursion

**Review by jamesr66a — COMMENTED** [review]
Looks pretty good to me. It could use some comments/docstrings describing what's going on at various places

File: torch/fx/experimental/gradual_graph_typechecker.py (hunk 3 above)
Comment (migeed-z) [reply torch/fx/experimental/gradual_graph_typechecker.py:50]: I thought more about this. The arguments are variables. We can either typecheck them by looking them up in the environment or we can typecheck them by calling the typechecker again which will go through the variable case and typecheck them. I can change it for now to just lookup in the type environment

**Pull request closed without merging**